### ISSUE TITLE
[Enterprise] Easier to follow changesets

### ISSUE THREAD

**Issue by davidr64** [rmosolgo/graphql-ruby#4267]
**Is your feature request related to a problem? Please describe.**

Our team has 17 changeset across our schema. When developers are working in an unfamiliar area for themselves it is not obvious when a changeset has modified a mutation or type being worked on. We can add comments around fields modified by changesets, however that is unreliable. We need a way to signal to developers a changeset has been applied to the object.

**Describe the solution you'd like**

I have two general ideas on a solution. 

1. Instead of defining changesets within separate classes we can either define a change_set function within the class we need to change or we import a changeset module into the type being changed that would define the changeset. This would at least make it obvious in the file that the type had been changed and we would be able to see successive changesets on a single object.  
2. We find a way to always have the newest behaviour defined in the main class file and use changesets to define the old behaviour. This would kind of be a reverse changeset where you access changesets to access older behaviour. This solution would still suffer from not knowing that the changeset exists however generally we are working the latest version of the schema not debugging old versions. The other benefit to this solution is it would allow us to refactor code when the changeset is made so that when the changeset is scheduled for removal it is as simple as deleting that file and no more code is required to move around when the changeset expires.

**Describe alternatives you've considered**

We can just add comments around code that is changed. This can be easily missed by the writer of the changesets.

**Comment by rmosolgo** [comment]
Hey, thanks for the detailed suggestion. I was just working with someone else who was using Changesets and found the same trouble, so I think we're on to something here. The good news is, I think we can accomplish an easier-to-follow definition API as "sugar" on top of the current implementation. Let me take a look over the next few days and I'll follow up back here!

**Comment by sirwolfgang** [comment]
I am attempting to do the same. Currently, my solution has been to extend types to accept a `release` and `deprecate` DSL that takes dates for when the type and related features should be shown, then removed.  

Then in my changesets, I will go back to the release version, and mark original fields, arguments, etc... and go to the effective change version and update it to match what's in my main file. 

```
def visible?(context)
  return false if release && context.api_version < release
  return false if deprecate && context.api_version > deprecate
  ...
end
```

**Comment by sirwolfgang** [comment]
Wanted to add, that another thing I did to make things easier to manage is change the file naming to match how DB migrations are, so it's `YYYY-MM-DD-feature_change.rb`. 

```ruby
# schema.rb
Dir['app/graphql/changesets/**/*.rb'].map { |filename| require_relative filename[12..-4] }

class Schema < GraphQL::Schema
  CHANGESETS = Dir['app/graphql/changesets/**/*.rb'].map do |filename|
    "Changesets::#{filename.split('/').last[11..-4].camelize}".constantize
  end

  use GraphQL::Enterprise::Changeset::Release, changesets: CHANGESETS
end
```

**Comment by davidr64** [comment]
@rmosolgo I am curious if you have more information on the " easier-to-follow definition API as "sugar" on top of the current implementation"

**Comment by dmcrorieGIT** [comment]
@rmosolgo Related to this, our organization is looking for an easier way to deprecate a Changeset completely. My team and I have been working on a way to do this, and after much discussion, there are two main solutions that we have come up with:
1. Have a file or module where we include all the deprecated changeset modifications, which will be applied first before all other changesets. The advantages to this are that we can easily see all of the history of all the changesets, and that would make automating deprecating a changeset version very easy (just copy and pasting). The con to this is that this probably won't scale well, and we would eventually have a massive file over time.
2. Add the modifications from an expired changeset directly into the schema. The pro to this is that this eliminates some bullky process, as the modifications are 99.9% of the time very straightforward to do. It would also look much cleaner compared to the first approach. The con is that this would probably be quite hard to automate (would need some advanced ruby metaprogramming logic to get this to work)

We are leaning heavily towards option two, since this seems to be the cleanest. Do you have any thoughts on the best possible approach to this problem?

**Comment by crpahl** [comment]
@rmosolgo Going to follow up on this one last time. Sorry for the noise!

Our schema is growing quickly and this is a change we'll need in the near future to make our changesets more manageable and our schema easier to follow as it continues to grow. We are dropping support for versions over time as they go out of our support window, but it's still a challenge due to our schema size.

So I wanted to follow up on this to make sure this isn't a thing that will be ready in the near future! If not, we'll develop something ourselves for now which shouldn't be a problem either!

**Comment by rmosolgo** [comment]
👋  Sure thing. Thanks everyone for sharing your experiences here and what you found that worked for you. I'm preparing a new release of the Changeset API to release early next week: 

- Add `changeset_dir: "..."` which requires each file in the named directory and attaches a changeset whose class name matches the filename (like @sirwolfgang's example in https://github.com/rmosolgo/graphql-ruby/issues/4267#issuecomment-1364203385)
- Add `added_in: ...` and `removed_in: ...` options to fields, arguments, enum values, possible types, and implements configurations so that changes can be configured _in_ their types instead of in separate changeset files. (The changeset would still be named and given a release date in a separate file, then referenced by name in the schema definition.) 

I think those would address several of the usability issues mentioned here -- please let me know what you think! Otherwise, I'll follow up here when that new API is released and documented.

**Comment by rmosolgo** [comment]
I just released `graphql-enterprise` v1.3.0 with a new `added_in: ...` / `removed_in: ...` API and updated the documentation to reflect it: https://graphql-ruby.org/changesets/definition.html

It uses all the same data structures under the hood and you should be able to migrate incrementally, one changeset (or even one _change_) at a time. If you give the new API a try and run into any problems, please let me know in a new issue!
